File: src/main/java/sprykerFileCreator/model/generator/ClassGenerator.java
package pav.sprykerFileCreator.model.generator;

import pav.sprykerFileCreator.model.definition.ClassDefinitionInterface;
import pav.sprykerFileCreator.model.definition.DefinitionProviderInterface;
import pav.sprykerFileCreator.model.renderer.dto.DocBlockItem;
import pav.sprykerFileCreator.model.renderer.dto.PhpClassInterface;
import pav.sprykerFileCreator.model.renderer.dto.PhpClassItem;

import java.util.HashMap;
import java.util.List;

public class ClassGenerator implements ClassGeneratorInterface {

    private ParentGeneratorInterface parentGenerator;
    private DocBlockGeneratorInterface docBlockGenerator;
    private DefinitionProviderInterface definitionProvider;

    public ClassGenerator(
            ParentGeneratorInterface parentGenerator,
            DocBlockGeneratorInterface docBlockGenerator,
            DefinitionProviderInterface definitionProvider
    ) {
        this.parentGenerator = parentGenerator;
        this.docBlockGenerator = docBlockGenerator;
        this.definitionProvider = definitionProvider;
    }

    @Override
    public PhpClassInterface generateClass(String classType, HashMap<String, String> config) throws Exception {

        ClassDefinitionInterface classDefinition = this.definitionProvider.getDefinitionByType(classType);

        String namespace = this.createNamespace(classDefinition, config);
        String className = this.createClassName(classDefinition, config);

        PhpClassInterface phpClass = new PhpClassItem(namespace, className);

        this.addParentClass(phpClass, classDefinition);
        this.addDocBlock(phpClass, classDefinition, config);

        return phpClass;
    }

    private String createNamespace(ClassDefinitionInterface classDefinition, HashMap<String, String> config) {

        String projectName = config.get(SprykerConstants.PROJECT_NAME);
        String bundleName = config.get(SprykerConstants.BUNDLE_NAME);

        return classDefinition.getNamespacePattern()
                .replace(SprykerConstants.PROJECT_NAME_PLACEHOLDER, projectName)
                .replace(SprykerConstants.BUNDLE_NAME_PLACEHOLDER, bundleName);
    }

    private String createClassName(ClassDefinitionInterface classDefinition, HashMap<String, String> config) {
        String className = "Index";
        String bundleName = config.get(SprykerConstants.BUNDLE_NAME);

        if (config.containsKey(SprykerConstants.CLASS_NAME)) {
            className = config.get(SprykerConstants.CLASS_NAME);
        }

        return classDefinition.getNamePattern()
                .replace(SprykerConstants.CLASS_NAME_PLACEHOLDER, className)
                .replace(SprykerConstants.BUNDLE_NAME_PLACEHOLDER, bundleName);
    }

    private void addParentClass(PhpClassInterface phpClass, ClassDefinitionInterface classDefinition) {
        PhpClassInterface parentClass = this.parentGenerator.getParentClass(phpClass, classDefinition);

        if (parentClass != null) {
            phpClass.setParentClass(parentClass);
        }
    }

    private void addDocBlock(
            PhpClassInterface phpClass,
            ClassDefinitionInterface classDefinition,
            HashMap<String, String> config
    ) throws Exception {
        String bundleName = config.get(SprykerConstants.BUNDLE_NAME);

        String[] docBlockClassTypes = classDefinition.getDocBlockClasses();
        List<DocBlockItem> docBlockItems =  this.docBlockGenerator.getDocBlockItems(docBlockClassTypes, bundleName);

        for (DocBlockItem docBlockItem: docBlockItems) {
            phpClass.addDocBlockItem(docBlockItem);
        }
    }
}

File: src/main/kotlin/com/github/tobi812/sprykerplugin/models/generator/ClassGenerator.kt
package com.github.tobi812.sprykerplugin.models.generator

import com.github.tobi812.sprykerplugin.actions.ClassConfig
import com.github.tobi812.sprykerplugin.constants.SprykerConstants
import com.github.tobi812.sprykerplugin.models.definitions.ClassDefinitionInterface
import com.github.tobi812.sprykerplugin.models.definitions.DefinitionProviderInterface
import com.github.tobi812.sprykerplugin.models.renderer.dto.DocBlockItem
import com.github.tobi812.sprykerplugin.models.renderer.dto.PhpClassInterface
import com.github.tobi812.sprykerplugin.models.renderer.dto.PhpClassItem
import java.lang.Exception

class ClassGenerator(
    private val parentGenerator: ParentGeneratorInterface,
    private val docBlockGenerator: DocBlockGeneratorInterface,
    private val definitionProvider: DefinitionProviderInterface
) : ClassGeneratorInterface {

    @Throws(Exception::class)
    override fun generateClass(classType: String, config: ClassConfig): PhpClassInterface {
        val classDefinition: ClassDefinitionInterface = this.definitionProvider.getDefinitionByType(classType)
        val namespace = this.createNamespace(classDefinition, config)
        val className = this.createClassName(classDefinition, config)
        val phpClass: PhpClassInterface = PhpClassItem(namespace, className)

        this.addParentClass(phpClass, classDefinition)
        this.addDocBlock(phpClass, classDefinition, config)

        return phpClass
    }

    override fun createFullQualifiedName(classDefinition: ClassDefinitionInterface, config: ClassConfig): String {
        val namespace: String = this.createNamespace(classDefinition, config)
        val className: String = this.createClassName(classDefinition, config)

        return namespace + "\\" + className
    }

    private fun createNamespace(classDefinition: ClassDefinitionInterface, config: ClassConfig): String {
        val projectName = config.projectName
        val moduleName = config.moduleName

        return classDefinition.namespacePattern
            .replace(SprykerConstants.PROJECT_NAME_PLACEHOLDER, projectName)
            .replace(SprykerConstants.MODULE_NAME_PLACEHOLDER, moduleName)
    }

    private fun createClassName(classDefinition: ClassDefinitionInterface, config: ClassConfig): String {
        val bundleName: String = config.moduleName
        val className: String = config.name ?: "Index"

        return classDefinition.namePattern
            .replace(SprykerConstants.CLASS_NAME_PLACEHOLDER, className)
            .replace(SprykerConstants.MODULE_NAME_PLACEHOLDER, bundleName)
    }

    private fun addParentClass(phpClass: PhpClassInterface, classDefinition: ClassDefinitionInterface) {
        val parentClass: PhpClassInterface? = parentGenerator.getParentClass(phpClass, classDefinition)
        if (parentClass != null) {
            phpClass.setParentClass(parentClass)
        }
    }

    @Throws(Exception::class)
    private fun addDocBlock(
        phpClass: PhpClassInterface,
        classDefinition: ClassDefinitionInterface,
        config: ClassConfig
    ) {
        val bundleName: String = config.moduleName
        val docBlockClassTypes: Array<String> = classDefinition.docBlockClasses
        val docBlockItems: List<DocBlockItem> = this.docBlockGenerator.getDocBlockItems(docBlockClassTypes, bundleName)

        for (docBlockItem in docBlockItems) {
            phpClass.addDocBlockItem(docBlockItem)
        }
    }
}
